pkg/controller/backupinfrastructure/backup_infrastructure_control.go

```diff
@@ -0,0 +1,338 @@
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+func (c *Controller) backupInfrastructureAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.backupInfrastructureQueue.Add(key)
+}
+
+func (c *Controller) backupInfrastructureUpdate(oldObj, newObj interface{}) {
+	var (
+		oldBackupInfrastructure = oldObj.(*gardenv1beta1.BackupInfrastructure)
+		newBackupInfrastructure = newObj.(*gardenv1beta1.BackupInfrastructure)
+		specChanged             = !apiequality.Semantic.DeepEqual(oldBackupInfrastructure.Spec, newBackupInfrastructure.Spec)
+		statusChanged           = !apiequality.Semantic.DeepEqual(oldBackupInfrastructure.Status, newBackupInfrastructure.Status)
+	)
+	if !specChanged && statusChanged {
+		return
+	}
+	c.backupInfrastructureAdd(newObj)
+}
+
+func (c *Controller) backupInfrastructureDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.backupInfrastructureQueue.Add(key)
+}
+
+func (c *Controller) reconcileBackupInfrastructureKey(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	backupInfrastructure, err := c.backupInfrastructureLister.BackupInfrastructures(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Debugf("[BACKUPINFRASTRUCTURE RECONCILE] %s - skipping because BackupInfrastructure has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[BACKUPINFRASTRUCTURE RECONCILE] %s - unable to retrieve object from store: %v", key, err)
+		return err
+	}
+
+	backupInfrastructureLogger := logger.NewFieldLogger(logger.Logger, "backupinfrastructure", fmt.Sprintf("%s/%s", backupInfrastructure.Namespace, backupInfrastructure.Name))
+
+	if backupInfrastructure.DeletionTimestamp != nil && !sets.NewString(backupInfrastructure.Finalizers...).Has(gardenv1beta1.GardenerName) {
+		backupInfrastructureLogger.Debug("Do not need to do anything as the BackupInfrastructure does not have my finalizer")
+		c.backupInfrastructureQueue.Forget(key)
+		return nil
+	}
+
+	durationToNextSync := c.config.Controllers.BackupInfrastructure.SyncPeriod.Duration
+	if reconcileErr := c.control.ReconcileBackupInfrastructure(backupInfrastructure, key); reconcileErr != nil {
+		durationToNextSync = 15 * time.Second
+	}
+	c.backupInfrastructureQueue.AddAfter(key, durationToNextSync)
+	backupInfrastructureLogger.Infof("Scheduled next reconciliation for BackupInfrastructure '%s' in %s", key, durationToNextSync)
```

Comment:
I thought of removing it earlier. But i think BackupInfrastructureController does significant task for checking bucket status on cloud provider compared to seed or secretbing etc. So, i kept it there. Specially in case of deletion we get to know when the next reconciliation will happen and it actually deletes it.  With default `deletionGracePeriodDays` set to 0, log will make more sense in my opinion.